src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -242,26 +255,83 @@ void AddReceivers(IBuilder builder, RecoverabilityExecutorFactory recoverability
             }
 
             foreach (var satellitePipeline in configuration.SatelliteDefinitions)
             {
                 var satelliteRecoverabilityExecutor = recoverabilityExecutorFactory.Create(satellitePipeline.RecoverabilityPolicy, satellitePipeline.ReceiveAddress);
                 var satellitePushSettings = new PushSettings(satellitePipeline.ReceiveAddress, errorQueue, configuration.PurgeOnStartup, satellitePipeline.RequiredTransportTransactionMode);
 
-                receivers.Add(new TransportReceiver(satellitePipeline.Name, messagePumpFactory(), satellitePushSettings, satellitePipeline.RuntimeSettings, new SatellitePipelineExecutor(builder, satellitePipeline), satelliteRecoverabilityExecutor, criticalError));
+                receivers.Add(new TransportReceiver(satellitePipeline.Name, messagePumpFactory, satellitePushSettings, satellitePipeline.RuntimeSettings, new SatellitePipelineExecutor(builder, satellitePipeline), satelliteRecoverabilityExecutor, criticalError, systemOutageConfiguration.RateLimitStartedNotification, systemOutageConfiguration.RateLimitEndedNotification));
             }
         }
 
+        async Task RateLimitLoop(CancellationToken cancellationToken)
+        {
+            //We want all the pumps to be running all the time in the desired mode until we call stop
+            //We want to make sure that StopRateLimit signal is not lost
+            //The circuit breaker ensures that if StartRateLimit has been called that eventually StopRateLimit is going to be called unless the endpoint stop
+
+            while (true)
+            {
+                await transitionBetweenRateLimitModes.WaitAsync(cancellationToken).ConfigureAwait(false);
+                var startRateLimiting = endpointShouldBeRateLimited;
+
+                while (true)
+                {
+                    try
+                    {
+                        foreach (var receiver in receivers)
+                        {
+                            if (startRateLimiting)
+                            {
+                                await receiver.StartRateLimiting().ConfigureAwait(false);
+                            }
+                            else
+                            {
+                                await receiver.StopRateLimiting().ConfigureAwait(false);
+                            }
+                        }
+
+                        break;
+                    }
+                    catch (Exception exception)
+                    {
+                        Logger.WarnFormat("Could not switch to {0} mode. '{1}'", startRateLimiting ? "rate limit" : "normal", exception);
+                        //Raise critical error
+                    }
+                }
+            }
+        }
+
+        void SwitchToRateLimitMode(Exception exception, long stateChangeTime)
+        {
+            if (stateChangeTime >= Interlocked.Read(ref lastStateChangeTime))
```

Comment:
Wouldn't it be simpler to set the bool flag and unleash the rate-limiting loop and let the loop figure out the rest? Then you also wouldn't need the state change time and you would be reading the lastStateChangeTime from a single location